autopush/router/apnsrouter.py

```diff
@@ -27,86 +28,175 @@ class APNSRouter(object):
               255: 'Unknown',
               }
 
-    def _connect(self):
-        """Connect to APNS"""
-        self.apns = apns.APNs(use_sandbox=self.config.get("sandbox", False),
-                              cert_file=self.config.get("cert_file"),
-                              key_file=self.config.get("key_file"),
-                              enhanced=True)
+    def _connect(self, cert_info):
+        """Connect to APNS
+
+        :param cert_info: APNS certificate configuration info
+        :type cert_info: dict
+
+        returns an instance of APNs that can be stored under the proper
+        release channel name.
+
+        """
+        # Do I still need to call this in _error?
+        return apns.APNs(
+            use_sandbox=cert_info.get("sandbox", False),
+            cert_file=cert_info.get("cert"),
+            key_file=cert_info.get("key"),
+            enhanced=True)
 
     def __init__(self, ap_settings, router_conf):
         """Create a new APNS router and connect to APNS"""
         self.ap_settings = ap_settings
         self._base_tags = []
-        self.config = router_conf
-        self.default_title = router_conf.get("default_title", "SimplePush")
-        self.default_body = router_conf.get("default_body", "New Alert")
-        self._connect()
-        self.log.debug("Starting APNS router...")
+        self.apns = dict()
+        self.messages = dict()
+        self._config = router_conf
+        if self._config.get('max_messages'):
+            self._max_messages = self._config['max_messages']
+            del(self._config['max_messages'])
+        else:
+            self._max_messages = 100
+        for rel_channel in self._config.keys():
+            self.apns[rel_channel] = self._connect(self._config[rel_channel])
+            self.apns[rel_channel].gateway_server.register_response_listener(
+                self._error)
+        self._connect(router_conf)
         self.ap_settings = ap_settings
+        # "unclean" is used by testing to prevent unclean reactor errors.
+        # It also prevents the local cleanup loop from running without
+        # requiring heavy lifting.
+        if "unclean" not in router_conf:
```

Comment:
same here, thus why I added this to just not call the reactor when we're running tests.